R/fair-aaa.R

```diff
@@ -0,0 +1,220 @@
+#'
+#' ```
+#' res_fairness <- new_groupwise_metric(...)
+#' res_by <- res_fairness(by)
+#' res_by(..., additional_arguments_to_.post = TRUE)
+#' ```
+#'
+#' For finer control of how groups in `by` are treated, use the
+#' `.post` argument.
+#'
+#' @return
+#' This function is a
+#' [function factory](https://adv-r.hadley.nz/function-factories.html); it's
+#' output is itself a function. Further, the functions that this function
+#' outputs are also function factories. More explicitly, this looks like:
+#'
+#' ```
+#' # a function with similar implementation to `demographic_parity()`:
+#' diff_range <- function(x) {diff(range(x$.estimate))}
+#'
+#' dem_parity <-
+#'   new_groupwise_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "dem_parity",
+#'     .post = diff_range
+#'   )
+#' ```
+#'
+#' The outputted `dem_parity` is a function that takes one argument, `by`,
+#' indicating the data-masked variable giving the sensitive feature.
+#'
+#' When called with a `by` argument, `dem_parity` will return a yardstick
+#' metric function like any other:
+#'
+#' ```
+#' dem_parity_by_gender <- dem_parity(gender)
+#' ```
+#'
+#' Note that `dem_parity` doesn't take any arguments other than `by`, and thus
+#' knows nothing about the data it will be applied to other than that it ought
+#' to have a column with name `"gender"` in it.
+#'
+#' The output `dem_parity_by_gender` is a metric function that takes the
+#' same arguments as the function supplied as `fn`, in this case
+#' `detection_prevalence`. It will thus interface like any other yardstick
+#' function except that it will look for a `"gender"` column in
+#' the data it's supplied.
+#'
+#' In addition to the examples below, see the documentation on the
+#' return value of fairness metrics like [demographic_parity()],
+#' [equal_opportunity()], or [equalized_odds()] to learn more about how the
+#' output of this function can be used.
+#'
+#' @examples
+#' data(hpc_cv)
+#'
+#' # `demographic_parity`, among other fairness metrics,
+#' # is generated with `new_groupwise_metric()`:
+#' diff_range <- function(x) {diff(range(x$.estimate))}
+#' demographic_parity_ <-
+#'   new_groupwise_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "demographic_parity",
+#'     .post = diff_range
+#'   )
+#'
+#' m_set <- metric_set(demographic_parity_(Resample))
+#'
+#' m_set(hpc_cv, truth = obs, estimate = pred)
+#'
+#' # the `post` argument can be used to accommodate a wide
+#' # variety of parameterizations. to encode demographic
+#' # parity as a ratio inside of a difference, for example:
+#' ratio_range <- function(x, ...) {
+#'   range <- range(x$.estimate)
+#'   range[1] / range[2]
+#' }
+#'
+#' demographic_parity_ratio <-
+#'   new_groupwise_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "demographic_parity_ratio",
+#'     .post = ratio_range
+#'   )
+#'
+#' @export
+new_groupwise_metric <- function(.fn, .name, .post, direction = "minimize") {
```

Comment:
The biggest remaining question for me is the naming of these arguments. Some stray thoughts:

* `.fn` was introduced for compatibility with `metric_tweak()`. Now that this function is named more similarly to `new_metric()`, and has a `direction` argument, maybe this should be `fn`?
* Davis suggested that `.post` could be `aggregate` or `.aggregate`. I do think "aggregate" or "summarize" is a more natural word for what's going on here, and at the same time am a little hesitant to name this argument in that way because they are functions and the tidy-style-aligned way to write about them would be `aggregate()` and `summarize()`, which suggest the actual stats and dplyr verbs to me.
* `.name` can switch to `name` if other arguments lose their `.` prefix.

Reply:
1. I think `fn` fits better, it aligns with `new_metric()` and the `*_metric_summarizer()` functions. Which makes sense, since this function is used to create metrics, not modify existing metrics which is what `metric_tweak()` is doing.
2. I would go with `aggregate`. We already use `summarize` in `collect_metrics()` and I think it would be confusing if it would be used differently here. `post`/`.post` might be more useful for post processing in workflows terminology.
3. Yes, switch to `name`. it will also align with `new_metric()` and the `*_metric_summarizer()` functions.